src/lib/transition_router/transition_router.ml

```diff
@@ -73,52 +73,66 @@ module Make (Inputs : Inputs_intf) :
     Strict_pipe.Reader.clear reader ;
     Strict_pipe.Writer.close writer
 
-  let is_transition_for_bootstrap ~frontier new_transition =
-    let root_transition =
-      Transition_frontier.root frontier
-      |> Transition_frontier.Breadcrumb.transition_with_hash |> With_hash.data
-    in
+  let is_transition_for_bootstrap root_state new_transition =
     let open External_transition in
-    let root_state = protocol_state (root_transition |> forget) in
     let new_state = protocol_state new_transition in
     Consensus.Mechanism.should_bootstrap
       ~existing:(Protocol_state.consensus_state root_state)
       ~candidate:(Protocol_state.consensus_state new_state)
 
-  let is_bootstrapping mode =
-    match mode with
-    | `Bootstrap_controller _ -> true
-    | `Transition_frontier_controller (_, _) -> false
+  let is_bootstrapping = function
+    | `Bootstrap_controller (_, _) -> true
+    | `Transition_frontier_controller (_, _, _) -> false
+
+  let get_root_state frontier =
+    Transition_frontier.root frontier
+    |> Transition_frontier.Breadcrumb.transition_with_hash |> With_hash.data
+    |> External_transition.forget |> External_transition.protocol_state
+
+  let set_bootstrap_mode ~frontier_ref ~mode root_state
+      bootstrap_controller_writer =
+    assert (not @@ is_bootstrapping !mode) ;
+    let _ : Transition_frontier.t = Mvar.take_now_exn frontier_ref in
+    mode := `Bootstrap_controller (root_state, bootstrap_controller_writer)
+
+  let set_transition_frontier_controller_mode ~frontier_ref ~mode new_frontier
+      reader writer =
+    assert (not @@ is_bootstrapping !mode) ;
+    assert (not @@ Mvar.is_empty frontier_ref) ;
+    Mvar.set frontier_ref new_frontier ;
+    mode := `Transition_frontier_controller (new_frontier, reader, writer)
```

Comment:
mode is an OCaml reference type that's being passed?

If it's global in scope, you could just make it a global within the module
If it needs to be delimited in scope, could you use a state monad?

Reply:
Hmmm... I am just going to inline to the inline function.